native/python/pyjp_buffer.cpp

```diff
@@ -0,0 +1,139 @@
+
+	sout << "<java beffer " << self->m_Buffer->getClass()->toString() << ">";
+	return JPPyString::fromStringUTF8(sout.str()).keep();
+	JP_PY_CATCH(0);
+}
+
+static void PyJPBuffer_releaseBuffer(PyJPBuffer *self, Py_buffer *view)
+{
+	JP_PY_TRY("PyJPBufferPrimitive_releaseBuffer");
+	JP_PY_CATCH();
+}
+
+int PyJPBuffer_getBuffer(PyJPBuffer *self, Py_buffer *view, int flags)
+{
+	JP_PY_TRY("PyJPBufferPrimitive_getBuffer");
+	JPContext *context = PyJPModule_getContext();
+	JPJavaFrame frame(context);
+	if (self->m_Buffer == NULL)
+		JP_RAISE(PyExc_ValueError, "Null buffer");
+	try
+	{
+		JPBuffer *buffer = self->m_Buffer;
+
+		if (!buffer->isValid())
+		{
+			PyErr_SetString(PyExc_BufferError, "Java buffer is not direct");
+			return -1;
+		}
+
+		if ((flags & PyBUF_WRITEABLE) == PyBUF_WRITEABLE && buffer->isReadOnly())
+		{
+			PyErr_SetString(PyExc_BufferError, "Java buffer is not writable");
+			return -1;
+		}
+
+		*view = buffer->getView();
+
+		// If strides are not requested and this is a slice then fail
+		if ((flags & PyBUF_STRIDES) != PyBUF_STRIDES)
+		{
+			if (view->strides[0] != view->itemsize)
+				JP_RAISE(PyExc_BufferError, "slices required strides");
+			view->strides = NULL;
+		}
+
+		// If shape is not requested
+		if ((flags & PyBUF_ND) != PyBUF_ND)
+		{
+			view->shape = NULL;
+		}
+
+		// If format is not requested
+		if ((flags & PyBUF_FORMAT) != PyBUF_FORMAT)
+			view->format = NULL;
+
+		// Okay all successful so reference the parent object
+		view->obj = (PyObject*) self;
+		Py_INCREF(view->obj);
+		return 0;
+	} catch (JPypeException &ex)
+	{
+		PyJPBuffer_releaseBuffer(self, view);
+
+		// We are only allowed to raise BufferError
+		PyErr_SetString(PyExc_BufferError, "Java buffer view failed");
+		return -1;
+	}
+	JP_PY_CATCH(-1);
+}
+
+static PyType_Slot bufferSlots[] = {
+	{ Py_tp_dealloc,  (void*) PyJPBuffer_dealloc},
+	{ Py_tp_repr,     (void*) PyJPBuffer_repr},
+	{0}
+};
+
+static PyBufferProcs directBuffer = {
+	(getbufferproc) & PyJPBuffer_getBuffer,
+	(releasebufferproc) & PyJPBuffer_releaseBuffer
+};
+
+PyTypeObject *PyJPBuffer_Type = NULL;
+static PyType_Spec bufferSpec = {
+	"_jpype._JBuffer",
+	sizeof (PyJPBuffer),
+	0,
+	Py_TPFLAGS_DEFAULT  | Py_TPFLAGS_BASETYPE,
+	bufferSlots
+};
+
+#ifdef __cplusplus
+}
+#endif
+
+void PyJPBuffer_initType(PyObject * module)
+{
+	JPPyTuple tuple = JPPyTuple::newTuple(1);
+	tuple.setItem(0, (PyObject*) PyJPObject_Type);
+	PyJPBuffer_Type = (PyTypeObject*) PyJPClass_FromSpecWithBases(&bufferSpec, tuple.get());
+	PyJPBuffer_Type->tp_as_buffer = &directBuffer;
+	JP_PY_CHECK_INIT();
+	PyModule_AddObject(module, "_JBuffer", (PyObject*) PyJPBuffer_Type);
+	JP_PY_CHECK_INIT();
+}
+
+JPPyObject PyJPBuffer_create(JPJavaFrame &frame, PyTypeObject *type, const JPValue& value)
+{
+	PyObject *obj = type->tp_alloc(type, 0);
+	JP_PY_CHECK();
+	((PyJPBuffer*) obj)->m_Buffer = new JPBuffer(value);
+	PyJPValue_assignJavaSlot(frame, obj, value);
+	return JPPyObject(JPPyRef::_claim, obj);
```

Comment:
this claims the reference to the buffer? So when it goes out of scope it will be garbage collected??!

Reply:
Okay this line likely need to be explained....

Normally we are using the ``_call`` or ``_use`` pattern  not the ``_claim``.   In a ``_use`` we are given an existing reference that we need to ensure we don't lose the reference to (like stupid things like manipulating a map and losing the reference to ourselves.   ``_call`` is also obvious because we got a reference an immediately put into a container.  

``_claim`` is an edge case.  We constructed a new reference, but for whatever reason we didn't put it in a reference counter (special error handling etc), but we want it to enter the reference counting system at some later point.

```
   JPPyObject doSomething()
   {

     // We created a reference but we can't throw because it requires a special handling...
    PyObject* obj = PyDo_Something(...);
     //  Next we do whatever checks we need and the handling routines...
    if (obj==NULL)
    {  // handle error
    }
     ...
    // Okay we know that it isn't 0 and but it still isn't in a reference counting frame so it
    // still has a reference count of 1.
   return JPPyObject(JPPyRef::_claim, obj);
   // So something weird happens here.  We claim the object so now it has a reference count 
   // of one and will die at the end of scope, but before that happens the return value will 
  // be constructed which will copy the reference and increment it to 2.  then decrement it 
  // back to 1.
}
```

So the intent of the line is to hold grab the reference and then transfer control to the return.   That said I do not see any reason for using the claim pattern in this case.  There is no special handling of types nor an exceptional return path.  And in some very rare case assignJavaSlot may fail resulting in a leaked object, though given the type system I am nor sure it can be triggered here.  I will convert it to a ``_call`` pattern to be safe.

Reply:
This code was revised and should be good to go.